packages/dds/tree/src/events/events.ts

```diff
@@ -97,26 +100,24 @@ export interface IEventEmitter<E extends Events<E>> {
  * Or, compose over it:
  * @example
  * ```ts
- * class MyClass extends IEventEmitter<MyEvents> {
+ * class MyClass implements IEventEmitter<MyEvents> {
  *   private readonly events = EventEmitter.create<MyEvents>();
  *
  *   private load() {
  *     this.events.emit("loaded");
  *   }
  *
- *   public on<K extends (string | symbol) & keyof MyEvents>(
+ *   public on<K extends keyof MyEvents>(
  *     eventName: K,
  *     listener: MyEvents[K],
  *   ): () => void {
- *     return events.on(eventName, listener);
+ *     return this.events.on(eventName, listener);
  *   }
  * }
  * ```
  */
 export class EventEmitter<E extends Events<E>> implements IEventEmitter<E> {
-    private readonly listeners: Partial<{
-        [P in keyof E]: Set<E[P]>;
-    }> = {};
+    private readonly listeners = new Map<keyof E, Set<(...args: unknown[]) => void>>();
 
     /**
      * Create an instance of an {@link EventEmitter}.
```

Comment:
Generally exporting non sealed classes is a bad idea, since semver compat for classes is kinda broken (ex: adding a private member is a breaking change).

I'd recommend moving this create function outside of the class (call it createEventEmitter).

Then we can package export this if we end up wanting to, without having to export the class.

Reply:
If you move the create function, then it can't use the type `EventEmitter<E>["emit"]` because it is protected.

Reply:
Besides, one of the usage patterns is to extend this class, so we do want to export it for that reason.

Reply:
This is no longer relevant after the latest changes.